### ISSUE TITLE
UFF optimization leads to overlapping atoms

### ISSUE THREAD

**Issue by oboril** [rdkit/rdkit#7901]
**Describe the bug**
When optimizing molecule conformers using UFF, sometimes the lowest energy conformer has 2 H atoms in (nearly) identical positions. The UFF energy for this conformer can be lower (~1kcal) compared to conformers with no overlapping atoms.

**To Reproduce**
```
from rdkit import Chem
from rdkit.Chem import AllChem

block = """\
Acetone, enolate form
     RDKit          3D

  9  8  0  0  0  0  0  0  0  0999 V2000
   -1.2143   -0.3494    0.0962 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.1072    0.3398    0.0041 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.1682    1.7287    0.0554 O   0  0  0  0  0  1  0  0  0  0  0  0
    1.2210   -0.3737   -0.1263 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.2311   -1.0203    0.9811 H   0  0  0  0  0  0  0  0  0  0  0  0
   -1.3951   -0.9468   -0.8225 H   0  0  0  0  0  0  0  0  0  0  0  0
   -2.0281    0.3987    0.2006 H   0  0  0  0  0  0  0  0  0  0  0  0
    2.1862    0.1115   -0.1943 H   0  0  0  0  0  0  0  0  0  0  0  0
    2.1861    0.1115   -0.1943 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0
  2  3  1  0
  2  4  2  0
  1  5  1  0
  1  6  1  0
  1  7  1  0
  4  8  1  0
  4  9  1  0
M  CHG  1   3  -1
M  END"""

# Load the mol block into an RDKit molecule
mol = Chem.MolFromMolBlock(block, removeHs=False)

# Calculate the UFF energy
energy = AllChem.UFFGetMoleculeForceField(mol).CalcEnergy()
print(energy)
# 1.9431066354329734

# optimize molecule with UFF
AllChem.UFFOptimizeMoleculeConfs(mol, maxIters=1000, ignoreInterfragInteractions=False)
energy = AllChem.UFFGetMoleculeForceField(mol).CalcEnergy()
print(energy)
# 1.9134088003298546
print(mol.GetConformer().GetPositions())
# [[-1.21431047 -0.34936551  0.09616025]
#  [ 0.10723219  0.33980251  0.00412303]
#  [ 0.168191    1.72867128  0.0552825 ]
#  [ 1.2209892  -0.37373454 -0.12628728]
#  [-1.23109677 -1.02040053  0.98105874]
#  [-1.39511005 -0.94675994 -0.82255978]
#  [-2.02805024  0.39873859  0.2007029 ]
#  [ 2.18613359  0.11151183 -0.19424122]
#  [ 2.18612156  0.11153631 -0.19423913]]

# Calculate the MMFF energy
energy = AllChem.MMFFGetMoleculeForceField(
    mol, AllChem.MMFFGetMoleculeProperties(mol)
).CalcEnergy()
print(energy)
# 211.6003579523681
```

I think this example shows the behaviour better than the random conformer generation.

Note that the last H atoms are in nearly identical positions, but UFF assigns very low energy (1.9 kcal/mol) and when asked to optimize the molecule, it places the H atoms even closer together.

MMFF correctly assigns this conformer high energy.

**Expected behavior**
UFF energy for conformers with overlapping atoms should be very high compared to conformers without overlapping atoms. MMFF seems to work correctly.

**Screenshots**

**Configuration (please complete the following information):**
 - RDKit version: 2024.3.5
 - OS: Ubuntu 24
 - Python version (if relevant): 3.12
 - Are you using conda? No
 - If you are using conda, which channel did you install the rdkit from?
 - If you are not using conda: how did you install the RDKit? using pip

**Additional context**

**Comment by oboril** [comment]
Looking into the nonbonded VdW force implementation:

```
double vdWContrib::getEnergy(double *pos) const {
  PRECONDITION(dp_forceField, "no owner");
  PRECONDITION(pos, "bad vector");

  double dist = dp_forceField->distance(d_at1Idx, d_at2Idx, pos);
  if (dist > d_thresh || dist <= 0.0) {
    return 0.0;
  }

  double r = d_xij / dist;
  double r6 = int_pow<6>(r);
  double r12 = r6 * r6;
  double res = d_wellDepth * (r12 - 2.0 * r6);
  // if(d_at1Idx==12 && d_at2Idx==21 ) std::cerr << "     >: " << d_at1Idx <<
  // "-" << d_at2Idx << " " << r << " = " << res << std::endl;
  return res;
}
```

The logic that if distance <= 0 then VdW energy is 0 seems like it could be causing the issue.

**Comment by greglandrum** [comment]
> The logic that if distance <= 0 then VdW energy is 0 seems like it could be causing the issue.

unfortunately without that logic, the calculation will generate a NaN and will crash.

In any case, here the distance should be > 0.0, but very small. That leads to very high forces and it might be that those are causing the problem.

**Comment by greglandrum** [comment]
I hadn't actually looked at the structure, so I was missing something. The two atoms that are very close to each other, atoms 7 and 8 in your structure, are actually both bonded to the same atom, atom 3. This means that the force field constructed by UFF has an angular term for the angle 7-3-8 and no vdw term between atoms 7 and 8.
I still need to do a bit of digging to look at the actual functional form for the angular term in order to figure out if something is wrong here or if this is just a pathology of UFF

**Comment by greglandrum** [comment]
Ok, to come back to this: the functional form that the UFF uses for angular terms around SP2 atoms has a minimum at 0 deg as well as the expected minima at 120 deg and 240 deg.  So since these two atoms start close to each other, they actually get pulled together instead of being pushed apart.
OpenBabel adds an extra term to their UFF implementation to deal with this. I will look into how we might do the same.

**Comment by oboril** [comment]
That would be great! I'm not very faminilar with FF functionals but I think it's reasonable to penalize overlapping atoms : )

**Comment by simonmb** [comment]
I can confirm this bug. I programmed a workaround to exclude conformers when this happens. But this is obviously no long term solution.

**Comment by simonmb** [comment]
@greglandrum, could the additional term be added?

**Comment by greglandrum** [comment]
@simonmb yeah, it's on my list to do this, it just ended up slipping too far from the top and I forgot about it.
I will try to get it done for the next release
